cache/src/main/java/io/envoyproxy/controlplane/cache/SimpleCache.java

```diff
@@ -1,195 +1,213 @@
+    T group = groups.hash(request.getNode());
 
-      // Trigger existing watches
-      Set<Long> idsToRemove = watches.row(group).entrySet().stream()
-          .map(watchEntry -> {
-            respond(watchEntry.getValue(), snapshot, group);
-            return watchEntry.getKey();
-          })
-          .collect(Collectors.toSet());
-
-      // Discard all watches; the client must request a new watch to receive updates and ACK/NACK.
-      //
-      // Note that the removals have to happen after we're done iterating over the row Map otherwise
-      // we get a java.util.ConcurrentModificationException
-      idsToRemove.forEach(idToRemove -> watches.remove(group, idToRemove));
-    }
-  }
-
-  /**
-   * {@inheritDoc}
-   */
-  @Override
-  public Watch watch(ResourceType type, Node node, String version, Collection<String> names) {
-    Watch watch = new Watch(names, type);
+    writeLock.lock();
 
-    final T group;
-
-    // Do nothing if group hashing failed.
     try {
-      group = groups.hash(node);
-    } catch (Exception e) {
-      LOGGER.error("failed to hash node {} into group", node, e);
+      CacheStatusInfo status = statuses.computeIfAbsent(group, g -> new CacheStatusInfo(request.getNode()));
 
-      return watch;
-    }
+      status.setLastWatchRequestTime(System.currentTimeMillis());
 
-    synchronized (lock) {
-      // If the requested version is up-to-date or missing a response, leave an open watch.
       Snapshot snapshot = snapshots.get(group);
+      String version = snapshot == null ? "" : snapshot.version(request.getTypeUrl());
 
-      if (snapshot == null || snapshot.version().equals(version)) {
-        if (snapshot == null && callback != null) {
-          LOGGER.info("callback {} at {}", group, version);
-
-          // TODO(jbratton): Should we track these CFs somewhere (e.g. to force completion on shutdown)?
-          executorService.submit(() -> callback.accept(group));
-        }
-
-        LOGGER.info("open watch for {}[{}] from key {} from version {}",
-            type,
-            String.join(", ", names),
-            group,
-            version);
+      Watch watch = new Watch(request);
 
+      // If the requested version is up-to-date or missing a response, leave an open watch.
+      if (snapshot == null || request.getVersionInfo().equals(version)) {
         watchCount++;
 
-        long id = watchCount;
+        long watchId = watchCount;
+
+        LOGGER.info("open watch {} for {}[{}] from node {} for version {}",
+            watchId,
+            request.getTypeUrl(),
+            String.join(", ", request.getResourceNamesList()),
+            group,
+            request.getVersionInfo());
 
-        watches.put(group, id, watch);
+        status.setWatch(watchId, watch);
 
         watch.setStop(() -> {
-          synchronized (lock) {
-            watches.remove(group, id);
+          writeLock.lock();
+
+          try {
+            status.removeWatch(watchId);
```

Comment:
same here. Status seems to have locking implementation of `removeWatch` also

Reply:
This is the same lock, but the nested lock statement here is executed at a later time because it is inside a separate func that's being passed to `setStop`.